File: python/neatxlsx/constant.py
from collections.abc import Mapping
from types import MappingProxyType
from typing import Literal, TypeAlias

from .spec import CellFormatPatch, XlsxWriteOptions

N_NROWS_EXCEL_MAX = 1_048_576
N_NCOLS_EXCEL_MAX = 16_384
N_LEN_EXCEL_SHEET_NAME_MAX = 31
TUP_EXCEL_ILLEGAL = ("*", ":", "?", "/", "\\", "[", "]")

# Strategy/Preference/Adjustable Parameters for XLSX I/O operations.

LIT_FMT_KEYS = Literal["text", "integer", "decimal", "scientific", "header"]
_cls_base_fmt_patch = CellFormatPatch(
    font_name="Times New Roman", font_size=11, border=1, align="left", valign="vcenter"
)

DEFAULT_XLSX_FORMATS: Mapping[LIT_FMT_KEYS, CellFormatPatch] = MappingProxyType(
    {
        "text": _cls_base_fmt_patch,
        "header": _cls_base_fmt_patch.with_(bold=True, align="center"),
        "integer": _cls_base_fmt_patch.with_(num_format="0"),
        "decimal": _cls_base_fmt_patch.with_(num_format="0.0000"),
        "scientific": _cls_base_fmt_patch.with_(num_format="0.00E+0"),
    }
)

DEFAULT_XLSX_WRITE_OPTIONS = XlsxWriteOptions()

# `str` means a literal column name; `int` means a zero-based column index.
ColumnIdentifier: TypeAlias = str | int

File: crates/neatxlsx_core/src/constant.rs
//! XLSX constants and default preset factories.

use std::collections::BTreeMap;

use crate::spec::{CellFormatPatch, XlsxWriteOptions};

/// Excel worksheet maximum row count.
pub const NROWS_SHEET_MAX: usize = 1_048_576;
/// Excel worksheet maximum column count.
pub const NCOLS_SHEET_MAX: usize = 16_384;
/// Excel sheet name maximum length.
pub const LEN_SHEET_NAME_MAX: usize = 31;
/// Characters not allowed in sheet names.
pub const SHEET_NAME_ILLEGAL_CHRS: [&str; 7] = ["*", ":", "?", "/", "\\", "[", "]"];

/// Canonical format preset keys.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FormatKey {
    /// Generic text cell format.
    Text,
    /// Integer number format.
    Integer,
    /// Decimal number format.
    Decimal,
    /// Scientific number format.
    Scientific,
    /// Header cell format.
    Header,
}

/// Column selector reference.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ColumnIdentifier {
    /// Select by column name.
    Name(String),
    /// Select by zero-based column index.
    Index(usize),
}

/// Build default named format presets used by [`crate::writer::XlsxWriter`].
pub fn create_default_xlsx_formats() -> BTreeMap<String, CellFormatPatch> {
    let base_format_spec = CellFormatPatch {
        font_name: Some("Times New Roman".to_string()),
        font_size: Some(11),
        border: Some(1),
        align: Some("left".to_string()),
        valign: Some("vcenter".to_string()),
        ..Default::default()
    };

    let mut formats = BTreeMap::new();
    formats.insert("text".to_string(), base_format_spec.clone());
    formats.insert(
        "header".to_string(),
        base_format_spec.with_(CellFormatPatch {
            bold: Some(true),
            align: Some("center".to_string()),
            ..Default::default()
        }),
    );
    formats.insert(
        "integer".to_string(),
        base_format_spec.with_(CellFormatPatch {
            num_format: Some("0".to_string()),
            ..Default::default()
        }),
    );
    formats.insert(
        "decimal".to_string(),
        base_format_spec.with_(CellFormatPatch {
            num_format: Some("0.0000".to_string()),
            ..Default::default()
        }),
    );
    formats.insert(
        "scientific".to_string(),
        base_format_spec.with_(CellFormatPatch {
            num_format: Some("0.00E+0".to_string()),
            ..Default::default()
        }),
    );

    formats
}

/// Build default write options.
pub fn create_default_xlsx_write_options() -> XlsxWriteOptions {
    XlsxWriteOptions::default()
}
